x/authenticator/types/authenticators.go

```diff
@@ -1,79 +1,170 @@
 package types
 
-// ToDo: consider moving to a different package
+// TODO: consider moving to a different package
 
 import (
+	fmt "fmt"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+
 	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
 )
 
 type Authenticator[T any] interface {
-	GetAuthenticationData(tx sdk.Tx, messageIndex uint8) T
-	Authenticate(ctx sdk.Context, msg sdk.Msg, authenticationData T) bool
+	GetAuthenticationData(tx sdk.Tx) (T, error)
+	Authenticate(ctx sdk.Context, authenticationData T, simulate bool) (err error)
```

Comment:
I don't think `simulate` should be part of the `Authenticate()` interface/signature. IMO, the best option here is to pass simulate as part of the `authenticationData`

Reply:
I also think we should have the message here and validate only the signature for that specific message. The caller will make sure this gets called for each message.

The reason for this is that in an authenticator-world, it is not certain that all messages in a tx will use the same authenticator (SigVerification).

I think a good solution here is to merge with the current validation, and then split it once we have the code in the ante handler that iterates and calls each authenticator individually